src/Microsoft.DotNet.Build.Tasks/PackageFiles/ApiCompat.targets

```diff
@@ -1,137 +1,121 @@
           Command="$(_ApiCompatCmd)"
-          CustomErrorRegularExpression="^[a-zA-Z]+ :" />
-  </Target>
+          CustomErrorRegularExpression="^[a-zA-Z]+ :"
+          StandardOutputImportance="Low"
+          IgnoreExitCode="true"
+    >
+      <Output TaskParameter="ExitCode" PropertyName="ApiCompatExitCode" />
+    </Exec>
 
-  <!-- Get the set of directories where the dependencies can be found -->
-  <Target Name="GetDependencyDirectories"
-          Returns="@(_DependencyDirectories)"
-          Condition="Exists('$(ReferenceProjectLockJson)')">
-    <ItemGroup>
-      <_DependencyDirectories Include="@(ReferencePath->'%(RootDir)%(Directory)')" />      
-    </ItemGroup>
-  </Target>
-  
-  <!-- We build the ref project and then get its dependencies -->
-  <Target Name="GetContractDependencies"
-          Returns="@(_ContractDependencies)"
-          Condition="Exists('$(ReferenceProjectLockJson)')">
-    <PrereleaseResolveNuGetPackageAssets
-                               AllowFallbackOnTargetSelection="true"
-                               IncludeFrameworkReferences="false"
-                               NuGetPackagesDirectory="$(PackagesDir)"
-                               ProjectLanguage="$(Language)"
-                               ProjectLockFile="$(ReferenceProjectLockJson)"
-                               TargetMonikers="$(NuGetTargetMoniker)">
-      <Output TaskParameter="ResolvedReferences" ItemName="ReferenceProjectReferences" />
-    </PrereleaseResolveNuGetPackageAssets>
-    <ItemGroup>
-      <_ContractDependencies Include="@(ReferenceProjectReferences->'%(RootDir)%(Directory)')" />      
-    </ItemGroup>    
+    <!--
+      To force incremental builds to show failures again we are invaliding
+       one compile input by touching the assembly info file
+    -->
+    <Touch Condition="'$(ApiCompatExitCode)'!='0'" Files="$(AssemblyInfoFile)" />
+    <Error Condition="'$(ApiCompatExitCode)'!='0'" Text="ApiCompat failed for '$(TargetPath)'" />
   </Target>
-  
+
   <!-- ApiCompat for Contract Assemblies -->
   <Target Name="ValidateApiCompatForRef"
-          AfterTargets="Build;CopyFilesToOutputDirectory"
-          DependsOnTargets="GetPreviousContractDependencyDirectories;GetPreviousContractAssembly;GetCurrentContractDependencies"
+          DependsOnTargets="CopyFilesToOutputDirectory"
           Condition="'$(RunApiCompatForRef)' == 'true' AND '$(RunApiCompat)' == 'true'" >
-    <PropertyGroup Condition="Exists('$(PreviousContractProjectLockJson)')">
-      <_ApiCompatCmd>"$(ToolsDir)CoreRun.exe" "$(ToolsDir)ApiCompat.exe" "$(_PreviousContractAssembly)"</_ApiCompatCmd>
+
+    <!--
+      This target is opportunistic in the sense it only runs if the previous contract version
+      has been built. If it doesn't find and older version then it will not run. This is because
+      we don't have a great way to always force that the older contract exists and has been built.
```

Comment:
Correct. We should consider serializing the build of the depproj and csproj files in the ref.builds in corefx to help with this.